### PR TITLE
Switch to ESModule

### LINKED ISSUE #221: Experiment: ESModule: A future without webpack
# Browser supporting
## Background
Chromium: ✔
Firefox: ✔
Firefox for Android: ❔
Firefox Geckoview: ❔
iOS: 👀 Can be done. Depends on `webextension-polyfill`

## Content 
Chromium: ✔
Firefox: ❌ https://bugzilla.mozilla.org/show_bug.cgi?id=1536094
Firefox for Android: ❌❔
Firefox Geckoview: ❌❔
iOS:  👀 Can be done. Depends on `webextension-polyfill`. Related https://github.com/Agoric/realms-shim/issues/47

## Options page:
Should not be a problem

## Pop up page:
Should not be a problem

# Toolchain supporting
📦 @pika/web https://github.com/pikapkg/web ❔ (not tested yet)
🔨 TypeScript custom transformation ✔

# Codebase that relies on Webpack
❔

### PR DESCRIPTION
**Pull request by Jack-Works**
close #221

# The new architecture of Maskbook

## TL;DR

With our new architecture, you can...

1. ✔️ use `import.meta`
1. ✔️ use dynamic import `import()` in any way you want
1. ✔️ still use `process.env.*`
1. ✔️ only start the minimal build tasks you need (e.g. `gulp watch-esm` or `gulp watch-assets`).
1. ✔️ easily to debug in the Chrome Devtools.
1. ✔️ use Hot Module Reload for React Components (experimental!!)
1. ✔️ still support `webext` toolchain
1. ❌ use `require()`
1. ❌ use all Webpack variables like `__webpack_public_path__`
1. ❌ tree shaking based on `process.env.NODE_ENV`

Caution about these things!

1. ⚠️ import a `JSON` file is supported, but not encouraged, please load the JSON file by `fetch` when possible.
1. ⚠️ When you change a `JSON` file that imported by `example.ts`, you need to change `example.ts` to recompile it because the JSON files are inlined to `example.ts`.
1. ⚠️ DO NOT publish the build on your machine to the store unless you remove the `temp` folder before the build because old files are not removed.
1. ⚠️ NEVER run `tsc` in this project! Run `ttsc` instead!
1. ⚠️ `webext` toolchain is not migrated to the new toolchain yet! That means you cannot automatically install this on Android. I'll try to add it later.

### Compile time

-   We now use `gulp` as our build tool. Use `gulp --tasks` to see what tasks are available.

    > Don't think `gulp` is an outdated tool! I deliberately select the correct tool based on the situation. For unbundled development, `gulp` is the best tool in the old days to handle multiple files without bundling them into a single file.

-   Source codes (`./src/`) are compiled by the TypeScript compiler directly to emit ES Module files.
-   All dependencies (`./node_modules/`) are compiled by Webpack and served as a global variable `__deps__`. (In future we will migrate some to `rollup` and `snowpack`).
-   Web Workers are compiled by Webpack.
-   The injected script is compiled by Webpack.
-   We use TypeScript incremental build to speed up build time.

### Runtime

-   In Chrome, everything is running in ES Module (besides injected script).
-   In Firefox, basically the same as Chrome besides content script. Content scripts are running in the System JS format.
    > Due to the bug ["Dynamic module import doesn't work in webextension content scripts"](https://bugzilla.mozilla.org/show_bug.cgi?id=1536094), it is impossible to use ES Module in the content script in Firefox.
-   In iOS, the [webextension polyfill for iOS](https://github.com/DimensionDev/webextension-polyfill) will provide the ability to load ES Module.

## To land this...

-   [x] Test it on Chrome
-   [x] Test it on Firefox
-   [x] Test it on Firefox for Android
-   [ ] Test it on iOS
-   [ ] Test it on GeckoView

## What's next?

-   Support React Dev tools
-   Better dependencies splitting
-   Investigate ES Module hot reload to improve DX.
-   Deprecates JSON import
-   Switch to rollup and Snowpack when possible
-   Move all image assets in the `./public/` folder to where they are be used and load them by `import.meta`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -116,36 +93,50 @@
...
+    "@types/uuid": "^8.0.0",
     "@types/webpack": "^4.41.18",
     "@typescript-eslint/eslint-plugin": "^3.4.0",
     "@typescript-eslint/parser": "^3.5.0",
-    "asmcrypto.js": "^2.3.2",
-    "awesome-typescript-loader": "^5.2.1",
-    "babel-loader": "^8.1.0",
     "chrome-launcher": "^0.13.3",
-    "copy-webpack-plugin": "^6.0.3",
+    "del": "^5.1.0",
```
Comment (septs) [inline package.json:139]: https://npm.im/rimraf

**Comment by Jack-Works** [comment]
@guanbinrui hi! can you help me to find why the E2E test failed?

**Comment by guanbinrui** [comment]
> @guanbinrui hi! can you help me to find why the E2E test failed?

Failed to open the browser. I think the `build/` folder is absent.

**Comment by Jack-Works** [comment]
@guanbinrui it's moved to temp/extension

**Comment by Jack-Works** [comment]
This PR is ready for review! @guanbinrui please help me to fix the E2E test thanks!

Also, please run this branch in the watch mode and try to modify `src/components/InjectedComponents/DecryptedPost/DecryptedPost.tsx` this component to see if Hot Module Reloading is looking good.

File: .storybook/main.js [hunk 2]
```diff
@@ -6,7 +6,7 @@ module.exports = {
         'storybook-addon-i18n/register.js',
         'storybook-addon-designs',
         '@storybook/addon-links',
-        '@storybook/addon-docs',
+        // '@storybook/addon-docs',
```
Comment (septs) [inline .storybook/main.js:9]: remove the comments

File: e2e/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+  "extends": "../tsconfig.json",
+  "include": ["./**/*"],
+  "compilerOptions": {}
```
Comment (septs) [inline e2e/tsconfig.json:4]: remove the line

File: jest.config.js [hunk 4]
```diff
@@ -15,10 +15,10 @@ module.exports = {
         require.resolve('fake-indexeddb/auto'),
         path.join(__dirname, './scripts/jest-setup.js'),
     ],
-    // skip packages other than 'ts-results' and 'holoflows/kit'
+    // skip packages other than 'ts-results'
     transformIgnorePatterns: [],
```
Comment (septs) [inline jest.config.js:19]: remove the line

File: package.json [hunk 5]
```diff
@@ -4,28 +4,25 @@
   "private": true,
   "license": "AGPL-3.0-or-later",
   "scripts": {
-    "build": "webpack --mode production",
-    "build-ci": "node --max_old_space_size=4096 -r ts-node/register/transpile-only scripts/ci-build.ts",
+    "build": "gulp build",
+    "build-ci": "node --max_old_space_size=4096 node_modules/gulp/bin/gulp.js ci",
```
Comment (septs) [inline package.json:8]:
```suggestion
    "build-ci": "gulp ci",
```

The problem should not exist anymore.

File: scripts/gulp/env.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    string: ['target', 'arch', 'resolution', 'firefox'],
+    boolean: ['reproducible'],
+    default: {
+        target: 'chromium',
+        arch: 'web',
+        resolution: 'desktop',
+        firefox: 'fennec',
+        reproducible: false,
+    },
+    alias: {
```
Comment (septs) [inline scripts/gulp/env.ts:11]: remove the `alias` field

File: scripts/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,5 @@
+{
+  "extends": "../tsconfig.json",
+  "include": ["./**/*"],
+  "compilerOptions": {}
```
Comment (septs) [inline scripts/tsconfig.json:4]: remove the line

File: src/env.d.ts [hunk 8]
```diff
@@ -2,18 +2,14 @@
...
-    readonly firefoxVariant: 'android' | 'desktop' | 'GeckoView' | undefined
-    readonly genericTarget: 'facebookApp' | 'browser'
-    readonly perferResponsiveTarget: 'xs' | undefined
-    readonly shadowRootMode: 'open' | 'closed'
-}
-
 declare module NodeJS {
     interface ProcessEnv {
         NODE_ENV: 'development' | 'production' | 'test'
         STORYBOOK?: boolean
```
Comment (septs) [inline src/env.d.ts:8]:
```suggestion
        STORYBOOK?: string
```

File: src/extension/options-page/DeveloperComponents/DecryptPost.tsx [hunk 9]
```diff
@@ -44,7 +44,7 @@ export function DecryptPostDeveloperMode() {
                     <DecryptPost
                         alreadySelectedPreviously={[]}
                         encryptedText={encryptedText}
-                        onDecrypted={(post) => {}}
+                        onDecrypted={() => {}}
```
Comment (septs) [inline src/extension/options-page/DeveloperComponents/DecryptPost.tsx:47]: use `_.noop`

File: src/tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,42 @@
+{
+  "extends": "../tsconfig.json",
+  // https://github.com/microsoft/TypeScript/issues/39700
+  // "include": [
```
Comment (septs) [inline src/tsconfig.json:4]: remove the comments

File: .storybook/main.js (hunk 2 above)
Comment (Jack-Works) [reply .storybook/main.js:9]: I want to add it in the future. This plugin installs babel register so gulp will use that instead of ts-node register. After we switch to pnpm, this problem will disappear cause we do not depend on babel register directly.

File: package.json (hunk 5 above)
Comment (Jack-Works) [reply package.json:8]: > The problem should not exist anymore.

Why?

File: src/env.d.ts (hunk 8 above)
Comment (Jack-Works) [reply src/env.d.ts:8]: it _is_ a boolean

File: .circleci/config.yml [hunk 11]
```diff
@@ -39,15 +39,12 @@ jobs:
...
-                  command: sudo apt-get install zip
             - run:
                   name: Build Maskbook
                   command: yarn build-ci
-            - store_artifacts:
-                  path: Maskbook.base.zip
-                  destination: /Maskbook.base.zip
+            # - store_artifacts:
+            #       path: Maskbook.base.zip
+            #       destination: /Maskbook.base.zip
```
Comment (guanbinrui) [inline .circleci/config.yml:47]:
```suggestion
```

File: .circleci/config.yml [hunk 12]
```diff
@@ -63,7 +60,7 @@ jobs:
             - persist_to_workspace:
                   root: ~/repo/
                   paths:
-                      - Maskbook.base.zip
+                      #   - Maskbook.base.zip
```
Comment (guanbinrui) [inline .circleci/config.yml:63]:
```suggestion
```

File: public/background.html [hunk 13]
```diff
@@ -0,0 +1,15 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <script src="/env.js"></script>
+        <script src="/polyfills/browser-polyfill.js"></script>
+        <script src="./loaders/dependencies.js"></script>
+        <script src="/patches/firefox-fix.js"></script>
+        <script src="./patches/permissions.js"></script>
+        <script src="/loaders/background-script.js"></script>
+        <meta charset="UTF-8" />
```
Comment (guanbinrui) [inline public/background.html:10]:
```suggestion
        <meta charset="UTF-8" />
        <script src="/env.js"></script>
        <script src="/polyfills/browser-polyfill.js"></script>
        <script src="./loaders/dependencies.js"></script>
        <script src="/patches/firefox-fix.js"></script>
        <script src="./patches/permissions.js"></script>
        <script src="/loaders/background-script.js"></script>
```

File: public/content-script.html [hunk 14]
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <meta charset="UTF-8" />
+        <script src="./loaders/dependencies.js"></script>
+        <script src="/polyfills/browser-polyfill.js"></script>
+        <script src="/patches/firefox-fix.js"></script>
+        <script src="/loaders/content-script.js"></script>
+
```
Comment (guanbinrui) [inline public/content-script.html:9]:
```suggestion
```

File: public/background.html (hunk 13 above)
Comment (Jack-Works) [reply public/background.html:10]: nice catch

File: package.json [hunk 15]
```diff
@@ -4,28 +4,25 @@
...
-    "build:ios": "run-s build:ios:webpack build:ios:ext:prebuilt",
-    "build:ios:ext:prebuilt": "ts-node scripts/webextensiom-shim-prebuilt.ts",
-    "build:ios:webpack": "webpack --mode production --wk-webview",
+    "build:base": "gulp build",
+    "build:chromium": "gulp build --chrome",
+    "build:e2e": "gulp build --e2e",
+    "build:firefox": "gulp build --fx",
+    "build:gecko": "gulp build --android",
+    "build:ios": "gulp build --ios",
     "compile:contracts": "./scripts/contracts.sh",
```
Comment (septs) [inline package.json:16]: use `gulp` impl?

Comment (Jack-Works) [reply package.json:16]: @guanbinrui?

File: src/extension/content-script/tasks.ts [hunk 16]
```diff
@@ -48,7 +48,7 @@ const _tasks = {
         (x) => x,
     ),
     async immersiveSetup(for_: ECKeyIdentifier) {
-        getActivatedUI().taskStartImmersiveSetup(for_)
+        ;(await getActivatedUI()).taskStartImmersiveSetup(for_)
```
Comment (guanbinrui) [inline src/extension/content-script/tasks.ts:51]:
```suggestion
        (await getActivatedUI()).taskStartImmersiveSetup(for_)
```

File: src/utils/type-transform/Payload.ts [hunk 17]
```diff
@@ -93,9 +92,10 @@ type Encoder = SocialNetworkWorkerAndUI['payloadEncoder']
 export function deconstructPayload(str: string, decoder: Decoder): Result<Payload, TypeError> {
     const decoders = (() => {
         if (isNil(decoder)) {
-            if (GetContext() === 'content') {
-                return [safeGetActiveUI().payloadDecoder]
-            }
+            // TODO:
```
Comment (guanbinrui) [inline src/utils/type-transform/Payload.ts:95]: TODO what?

File: src/utils/type-transform/SECP256k1-Compression.ts [hunk 18]
```diff
@@ -1,6 +1,7 @@
 import secp256k1 from 'tiny-secp256k1'
 import { Convert, combine } from 'pvtsutils'
 import { encodeArrayBuffer, decodeArrayBuffer } from './String-ArrayBuffer'
+import { Buffer } from 'buffer'
```
Comment (guanbinrui) [inline src/utils/type-transform/SECP256k1-Compression.ts:4]: Why not `ArrayBuffer`?

File: src/extension/content-script/tasks.ts [hunk 19]
```diff
@@ -48,7 +48,7 @@ const _tasks = {
         (x) => x,
     ),
     async immersiveSetup(for_: ECKeyIdentifier) {
```
Comment (septs) [inline src/extension/content-script/tasks.ts:50]:
```suggestion
    async immersiveSetup(identifier: ECKeyIdentifier) {
```

Comment (septs) [inline src/extension/content-script/tasks.ts:50]:
```suggestion
    async immersiveSetup(identifier: ECKeyIdentifier) {
```

File: src/utils/type-transform/SECP256k1-Compression.ts (hunk 18 above)
Comment (Jack-Works) [reply src/utils/type-transform/SECP256k1-Compression.ts:4]: The original code is using `Buffer` global without referencing it. It breaks under our new architecture. We may change them later

File: package.json (hunk 15 above)
Comment (septs) [reply package.json:16]: the task not migrate?

Comment (Jack-Works) [reply package.json:16]: leave that in future maybe. it's not a frequent op and is isolated to other tasks

File: scripts/gulp/web-ext.d.ts [hunk 20]
```diff
@@ -0,0 +1,682 @@
+/**
+ * WARNING: THIS FILE IS GENERATED. DO NOT MODIFY IT.
+ * https://github.com/mozilla/web-ext/issues/1986
+ */
+declare module 'web-ext' {
```
Comment (septs) [inline scripts/gulp/web-ext.d.ts:5]: extract to a common library?

```
@dimensiondev/types
```
